lib/awful/completion.lua

```diff
@@ -100,15 +100,17 @@ function completion.shell(command, cur_pos, ncomp, shell)
             -- NOTE: ${~:-"..."} turns on GLOB_SUBST, useful for expansion of
             -- "~/" ($HOME).  ${:-"foo"} is the string "foo" as var.
             shell_cmd = "/usr/bin/env zsh -c 'local -a res; res=( ${~:-"
-                .. string.format('%q', words[cword_index]) .. "}* ); "
+                .. string.format('%q', words[cword_index]) .. "}*(N) ); "
                 .. "print -ln -- ${res[@]}'"
         else
-            -- check commands, aliases, builtins, functions and reswords
-            shell_cmd = "/usr/bin/env zsh -c 'local -a res; "..
+            -- Check commands, aliases, builtins, functions and reswords.
+            -- Adds executables and non-empty dirs from $PWD (pwd_exe).
+            shell_cmd = "/usr/bin/env zsh -c 'local -a res pwd_exe; "..
+            "pwd_exe=(*(N*:t) *(NF:t)); "..
             "res=( "..
             "\"${(k)commands[@]}\" \"${(k)aliases[@]}\" \"${(k)builtins[@]}\" \"${(k)functions[@]}\" "..
             "\"${(k)reswords[@]}\" "..
-            "${PWD}/*(:t)"..
+            "./${^${pwd_exe}} "..
```

Comment:
> I've looked into adding tests, but there does not seem to be an easy way to change the cwd in busted tests, is there?

Since (by that other comment)  you seem to be mocking `io.popen` now, I guess that this is no longer required? Otherwise: Nope, no easy way. You could use `lgi.GLib.chdir("/tmp")`. That would allow splitting things into two parts: One where `popen` is mocked and produces "known good" results and another part where we test that zsh produces the output we want from it. For this it might make sense to change the directory in `before_test(function() end)` and change back in `after_test()` (`lgi.GLib.get_current_dir()`).